<a href="https://colab.research.google.com/github/Thitichaya2000/DWDM21/blob/main/Project_lookmhee.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**ปัญหาคือ**
ไม่สามารถพยากรณ์ข้อมูลของปีต่อๆไปได้เนื่องจากไม่ทราบความสัมพันธ์ของข้อมูลปัจจุบันที่มีอยู่

**วัตถุประสงค์ คือ** ศึกษาความสัมพันธ์ของตารางที่ 1-4 ว่าความสัมพันธ์ที่ได้จากการคาดเดาปีต่อๆไปเป็นอย่างไร?



In [1]:
import pandas as pd   
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path = '/content/drive/My Drive/PROJECT_DWDM'

#ข้อมูลชุดที่ 1 จำนวนผู้ลงทะเบียนใช้บริการอินเทอร์เน็ตความเร็วสูง 
https://data.go.th/dataset/broadband-subscribers?fbclid=IwAR2hJLTNGiEnMNwRApUdbzzaFUj-1_Ag7HWgdZPxX-oaEKw2PBU0_W3-_m0


คำอธิบายตาราง

no. คือ ลำดับ

quarter คือ ไตรมาส (4 ไตรมาส)

Thailand Internet Users_subscriber คือ ประเภท year คือ ปี

value_subscriber คือ จำนวนผู้ลงทะเบียนใช้บริการอินเทอร์เน็ตความเร็วสูง

In [4]:
data_subscriber = pd.read_csv(os.path.join(path,'10broadband-subscribers.csv'))  #จำนวนผู้ลงทะเบียนใช้บริการอินเทอร์เน็ตความเร็วสูงเก็บไว้ในตัวแปร data_subscriber
data_subscriber

no.  quarter Thailand Internet Users_subscriber  year  value_subscriber
0     4        4         Total Internet subscribers  2003         652726.00
1     3        3         Total Internet subscribers  2003         529530.00
2     2        2         Total Internet subscribers  2003         453043.00
3     1        1         Total Internet subscribers  2003         365219.00
4     8        4         Total Internet subscribers  2004        1231344.00
..  ...      ...                                ...   ...               ...
68   72        4         Total Internet subscribers  2020       11478264.92
69   71        3         Total Internet subscribers  2020       11282645.58
70   70        2         Total Internet subscribers  2020       10912996.14
71   69        1         Total Internet subscribers  2020       10264995.67
72   73        1         Total Internet subscribers  2021       11876158.44

[73 rows x 5 columns]

In [5]:
data_subscriber2 = data_subscriber.sort_values(by=['no.'])   #เรียงข้อมูลโดยใช้ No. เก็บในตัวแปร data_subscriber2
data_subscriber2

no.  quarter Thailand Internet Users_subscriber  year  value_subscriber
3     1        1         Total Internet subscribers  2003         365219.00
2     2        2         Total Internet subscribers  2003         453043.00
1     3        3         Total Internet subscribers  2003         529530.00
0     4        4         Total Internet subscribers  2003         652726.00
7     5        1         Total Internet subscribers  2004         785931.00
..  ...      ...                                ...   ...               ...
71   69        1         Total Internet subscribers  2020       10264995.67
70   70        2         Total Internet subscribers  2020       10912996.14
69   71        3         Total Internet subscribers  2020       11282645.58
68   72        4         Total Internet subscribers  2020       11478264.92
72   73        1         Total Internet subscribers  2021       11876158.44

[73 rows x 5 columns]

#ข้อมูลชุดที่ 2 อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนครัวเรือน
https://data.go.th/dataset/broadband-penetration-per-household?fbclid=IwAR0tNweDa7pmiAaygo0Z88iFFIYLtQRFjujfaIyEpUub_bFVQEkcCbXAZJc


คำอธิบายตาราง

no. คือ ลำดับ

type_household คือ ประเภท

quarter คือ ไตรมาส (4 ไตรมาส)

year คือ ปี

value_household คือ อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนครัวเรือน

In [6]:
data_household = pd.read_csv(os.path.join(path,'12broadband-penetration-per-household-.csv'))  #อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนครัวเรือนเก็บไว้ในตัวแปร data_household
data_household

no.     type_household  quarter  year  value_household
0     4  per household (%)        4  2003         0.000649
1     3  per household (%)        3  2003         0.000459
2     2  per household (%)        2  2003         0.000343
3     1  per household (%)        1  2003         0.000281
4     8  per household (%)        4  2004         0.008716
..  ...                ...      ...   ...              ...
68   72  per household (%)        4  2020         0.514683
69   71  per household (%)        3  2020         0.515873
70   70  per household (%)        2  2020         0.498972
71   69  per household (%)        1  2020         0.469344
72   73  per household (%)        1  2021         0.532524

[73 rows x 5 columns]

In [7]:
data_household2 = data_household.sort_values(by=['no.'])  #เรียงข้อมูลโดยใช้ No. เก็บในตัวแปร data_household2
data_household2

no.     type_household  quarter  year  value_household
3     1  per household (%)        1  2003         0.000281
2     2  per household (%)        2  2003         0.000343
1     3  per household (%)        3  2003         0.000459
0     4  per household (%)        4  2003         0.000649
7     5  per household (%)        1  2004         0.001556
..  ...                ...      ...   ...              ...
71   69  per household (%)        1  2020         0.469344
70   70  per household (%)        2  2020         0.498972
69   71  per household (%)        3  2020         0.515873
68   72  per household (%)        4  2020         0.514683
72   73  per household (%)        1  2021         0.532524

[73 rows x 5 columns]

#ข้อมูลชุดที่ 3 อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนประชากร

https://data.go.th/dataset/broadband-penetration-per-population?fbclid=IwAR0D5Z4yxZOXJYPM9hgBJpAVbuWBkxwNzEdW_jWwlvvkCKQxaGERqrxffFs


คำอธิบายตาราง

no. คือ ลำดับ

type_population คือ ประเภท

quarter คือ ไตรมาส (4 ไตรมาส)

year คือ ปี

value_population คือ อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนประชากร

In [8]:
data_population = pd.read_csv(os.path.join(path,'11broadband-penetration-per-population-.csv'))  #อัตราการเข้าถึงของบริการอินเทอร์เน็ตความเร็วสูงต่อจำนวนประชากร ชื่อ data_population
data_population

no.     type_population  quarter  year  value_population
0     1  per population (%)        1  2003            0.0001
1     2  per population (%)        2  2003            0.0001
2     3  per population (%)        3  2003            0.0001
3     4  per population (%)        4  2003            0.0002
4     5  per population (%)        1  2004            0.0005
..  ...                 ...      ...   ...               ...
68   69  per population (%)        1  2020            0.1510
69   70  per population (%)        2  2020            0.1605
70   71  per population (%)        3  2020            0.1659
71   72  per population (%)        4  2020            0.1685
72   73  per population (%)        1  2021            0.1740

[73 rows x 5 columns]

###เชค Missing

In [9]:
data_household2.isnull().any()  #check missing ตาราง data_household2 ขึ้น False คือไม่มี Missing 

no.                False
type_household     False
quarter            False
year               False
value_household    False
dtype: bool

In [10]:
data_subscriber2.isnull().any()  #check missing ตาราง data_subscriber2

no.                                   False
quarter                               False
Thailand Internet Users_subscriber    False
year                                  False
value_subscriber                      False
dtype: bool

In [11]:
data_population.isnull().any()  #check missing ตาราง data_population

no.                 False
type_population     False
quarter             False
year                False
value_population    False
dtype: bool

####ต่อตาราง

In [12]:
data_population.drop(['quarter','year'], axis=1, inplace=True)  #drop colunm ไตรมาสกับปีออกเพราะชุดข้อมูลที่จะเชื่อมมีเหมือนกัน
data_population

no.     type_population  value_population
0     1  per population (%)            0.0001
1     2  per population (%)            0.0001
2     3  per population (%)            0.0001
3     4  per population (%)            0.0002
4     5  per population (%)            0.0005
..  ...                 ...               ...
68   69  per population (%)            0.1510
69   70  per population (%)            0.1605
70   71  per population (%)            0.1659
71   72  per population (%)            0.1685
72   73  per population (%)            0.1740

[73 rows x 3 columns]

In [13]:
merge_table = data_subscriber2.merge(data_population,how='left', left_on='no.',right_on='no.')  # เชื่อมตาราง data_subscriber2 กับ data_population  โดยให้ตาราง data_subscriber2 เป็นตารางหลักอยู่ที่ด้านซ้าย
merge_table

no.  quarter  ...     type_population  value_population
0     1        1  ...  per population (%)            0.0001
1     2        2  ...  per population (%)            0.0001
2     3        3  ...  per population (%)            0.0001
3     4        4  ...  per population (%)            0.0002
4     5        1  ...  per population (%)            0.0005
..  ...      ...  ...                 ...               ...
68   69        1  ...  per population (%)            0.1510
69   70        2  ...  per population (%)            0.1605
70   71        3  ...  per population (%)            0.1659
71   72        4  ...  per population (%)            0.1685
72   73        1  ...  per population (%)            0.1740

[73 rows x 7 columns]

In [14]:
data_household2.drop(['quarter','year'], axis=1, inplace=True)  #drop ปีกับไตรมาสออก
data_household2

no.     type_household  value_household
3     1  per household (%)         0.000281
2     2  per household (%)         0.000343
1     3  per household (%)         0.000459
0     4  per household (%)         0.000649
7     5  per household (%)         0.001556
..  ...                ...              ...
71   69  per household (%)         0.469344
70   70  per household (%)         0.498972
69   71  per household (%)         0.515873
68   72  per household (%)         0.514683
72   73  per household (%)         0.532524

[73 rows x 3 columns]

In [15]:
merge_table2 = merge_table.merge(data_household2,how='left', left_on='no.',right_on='no.')  # เชื่อมตาราง merge_table และ data_household2  โดยให้ตาราง merge_table เป็นตารางหลักอยู่ที่ด้านซ้าย
merge_table2

no.  quarter  ...     type_household  value_household
0     1        1  ...  per household (%)         0.000281
1     2        2  ...  per household (%)         0.000343
2     3        3  ...  per household (%)         0.000459
3     4        4  ...  per household (%)         0.000649
4     5        1  ...  per household (%)         0.001556
..  ...      ...  ...                ...              ...
68   69        1  ...  per household (%)         0.469344
69   70        2  ...  per household (%)         0.498972
70   71        3  ...  per household (%)         0.515873
71   72        4  ...  per household (%)         0.514683
72   73        1  ...  per household (%)         0.532524

[73 rows x 9 columns]

#ชุดข้อมูลที่อาจารย์ให้หาเพิ่ม  จำนวนเลขหมายโทรศัพท์เคลื่อนที่ที่ได้รับการจดทะเบียน
https://data.go.th/dataset/mobile-subscribers?fbclid=IwAR3ZoY2gqT_ogfZwHHI40nC_ejcmi3O9xZMOmcac_rCCdA5brXcHVh85Ayw

In [16]:
data_mobile = pd.read_csv(os.path.join(path,'05mobile-subscribers.csv'))  #จำนวนเลขหมายโทรศัพท์เคลื่อนที่ที่ได้รับการจดทะเบียน เก็บในตัวแปรชื่อdata_mobile
data_mobile

no. Mobile Market Report  quarter  year  value_mobile
0     4   Mobile Subscribers        4  2002      17449890
1     3   Mobile Subscribers        3  2002      15743776
2     2   Mobile Subscribers        2  2002      12416261
3     1   Mobile Subscribers        1  2002       9669909
4     8   Mobile Subscribers        4  2003      21616910
..  ...                  ...      ...   ...           ...
72   76   Mobile Subscribers        4  2020     116294420
73   75   Mobile Subscribers        3  2020     119169759
74   74   Mobile Subscribers        2  2020     131881072
75   73   Mobile Subscribers        1  2020     132594791
76   77   Mobile Subscribers        1  2021     117562201

[77 rows x 5 columns]

In [17]:
data_mobile2 = data_mobile.sort_values(by=['no.'])  #เรียงข้อมูลตาม No.
data_mobile2

no. Mobile Market Report  quarter  year  value_mobile
3     1   Mobile Subscribers        1  2002       9669909
2     2   Mobile Subscribers        2  2002      12416261
1     3   Mobile Subscribers        3  2002      15743776
0     4   Mobile Subscribers        4  2002      17449890
7     5   Mobile Subscribers        1  2003      18763102
..  ...                  ...      ...   ...           ...
75   73   Mobile Subscribers        1  2020     132594791
74   74   Mobile Subscribers        2  2020     131881072
73   75   Mobile Subscribers        3  2020     119169759
72   76   Mobile Subscribers        4  2020     116294420
76   77   Mobile Subscribers        1  2021     117562201

[77 rows x 5 columns]

In [18]:
data_mobile2.drop(['quarter','year'], axis=1, inplace=True)  #drop ปีและไตรมาสออกเพราะคอลัมน์ซ้ำ
data_mobile2

no. Mobile Market Report  value_mobile
3     1   Mobile Subscribers       9669909
2     2   Mobile Subscribers      12416261
1     3   Mobile Subscribers      15743776
0     4   Mobile Subscribers      17449890
7     5   Mobile Subscribers      18763102
..  ...                  ...           ...
75   73   Mobile Subscribers     132594791
74   74   Mobile Subscribers     131881072
73   75   Mobile Subscribers     119169759
72   76   Mobile Subscribers     116294420
76   77   Mobile Subscribers     117562201

[77 rows x 3 columns]

In [19]:
data_mobile3 = data_mobile2.iloc[4:,:]  #ตัดข้อมูล 4 แถวแรกออกเพราะมันเป็นของปี 2002 แต่ตาราง merge_table2 มันเริ่มที่ปี 2003
data_mobile3

no. Mobile Market Report  value_mobile
7     5   Mobile Subscribers      18763102
6     6   Mobile Subscribers      19788956
5     7   Mobile Subscribers      20655866
4     8   Mobile Subscribers      21616910
11    9   Mobile Subscribers      23217311
..  ...                  ...           ...
75   73   Mobile Subscribers     132594791
74   74   Mobile Subscribers     131881072
73   75   Mobile Subscribers     119169759
72   76   Mobile Subscribers     116294420
76   77   Mobile Subscribers     117562201

[73 rows x 3 columns]

In [20]:
merge_table3 = data_mobile3.merge(merge_table2,how='left', left_on='no.',right_on='no.')  # เชื่อมตาราง merge_table2 และ data_mobile เโดยให้ตาราง merge_table2 เป็นตารางหลักอยู่ที่ด้านซ้าย
merge_table3

no. Mobile Market Report  ...     type_household  value_household
0     5   Mobile Subscribers  ...  per household (%)         0.001556
1     6   Mobile Subscribers  ...  per household (%)         0.002265
2     7   Mobile Subscribers  ...  per household (%)         0.004964
3     8   Mobile Subscribers  ...  per household (%)         0.008716
4     9   Mobile Subscribers  ...  per household (%)         0.015449
..  ...                  ...  ...                ...              ...
68   73   Mobile Subscribers  ...  per household (%)         0.532524
69   74   Mobile Subscribers  ...                NaN              NaN
70   75   Mobile Subscribers  ...                NaN              NaN
71   76   Mobile Subscribers  ...                NaN              NaN
72   77   Mobile Subscribers  ...                NaN              NaN

[73 rows x 11 columns]

In [21]:
merge_table3.isnull().any()  #เชคค่า Missing พบว่าขึ้นมีขึ้น  True นั่นคือมันมี Missing

no.                                   False
Mobile Market Report                  False
value_mobile                          False
quarter                                True
Thailand Internet Users_subscriber     True
year                                   True
value_subscriber                       True
type_population                        True
value_population                       True
type_household                         True
value_household                        True
dtype: bool

In [22]:
merge_table3_drop = merge_table3.dropna()  #จัดการกับ Missing ด้วยการ dropna
merge_table3_drop

no. Mobile Market Report  ...     type_household  value_household
0     5   Mobile Subscribers  ...  per household (%)         0.001556
1     6   Mobile Subscribers  ...  per household (%)         0.002265
2     7   Mobile Subscribers  ...  per household (%)         0.004964
3     8   Mobile Subscribers  ...  per household (%)         0.008716
4     9   Mobile Subscribers  ...  per household (%)         0.015449
..  ...                  ...  ...                ...              ...
64   69   Mobile Subscribers  ...  per household (%)         0.469344
65   70   Mobile Subscribers  ...  per household (%)         0.498972
66   71   Mobile Subscribers  ...  per household (%)         0.515873
67   72   Mobile Subscribers  ...  per household (%)         0.514683
68   73   Mobile Subscribers  ...  per household (%)         0.532524

[69 rows x 11 columns]

In [23]:
merge_table3_drop.isnull().any() #เชค missing อีกรอบพบว่าไม่มีแล้ว

no.                                   False
Mobile Market Report                  False
value_mobile                          False
quarter                               False
Thailand Internet Users_subscriber    False
year                                  False
value_subscriber                      False
type_population                       False
value_population                      False
type_household                        False
value_household                       False
dtype: bool

In [24]:
removed = merge_table3.shape[0] - merge_table3_drop.shape[0]  #เชคว่าข้อมูลหายไปกี% จากการ dropna

print(f'size before drop = {merge_table3.shape[0]}')
print(f'size aftre drop = {merge_table3_drop.shape[0]}')
print(f'we loss {100*(removed/merge_table3.shape[0])}% of data')

size before drop = 73
size aftre drop = 69
we loss 5.47945205479452% of data


พบว่าจากการ dropna ข้อมูลหาย 5.47945205479452% ซึ่งยอมรับได้

###กำหนดค่าในตาราง เพื่อนำข้อมูลในตารางไปใช้ในการ Assosiation

โดยการเพิ่ม column ใหม่ในตาราง merge_table3_drop และ column ที่สร้างใหม่นั้นจะตามหลังด้วย _group โดยการจัดกลุ่มให้ค่าทั้งหมดในแต่ละ column ซึ่งจัดกลุ่มดังนี้

วัตถุประสงค์ คือ ศึกษาความสัมพันธ์ของตารางที่ 1-4 ว่าความสัมพันธ์ที่ได้จากการคาดเดาปีต่อๆไปเป็นยังไง?

In [25]:
from pandas.api.types import CategoricalDtype

###เพิ่ม column ใหม่ในตาราง merge_table3_drop

In [26]:
merge_table3_drop.columns

Index(['no.', 'Mobile Market Report', 'value_mobile', 'quarter',
       'Thailand Internet Users_subscriber', 'year', 'value_subscriber',
       'type_population', 'value_population', 'type_household',
       'value_household'],
      dtype='object')

column ชื่อว่า value_subscriber

ถ้ามีค่าตั้งแต่ 0 - 2000000 จะให้เป็น low_subscriber

ถ้ามีค่าตั้งแต่ 2000000 - 5000000 จะให้เป็น median_subscriber

ถ้ามีค่าตั้งแต่ 5000000 - 12000000 จะให้เป็น high_subscriber

In [43]:
bucket = ['low_subscriber','median_subscriber','high_subscriber']
merge_table3_drop['subscriber_group'] = pd.cut(merge_table3_drop.value_subscriber,
                            bins=[0,2000000,5000000,12000000],
                            labels=bucket,
                            right=False).astype(str).astype(CategoricalDtype(bucket,ordered=True))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


column ชื่อว่า value_household

ถ้ามีค่าตั้งแต่ 0 - 0.1 จะให้เป็น low_household

ถ้ามีค่าตั้งแต่ 0.1 - 0.3 จะให้เป็น median_household

ถ้ามีค่าตั้งแต่ 0.3 - 0.6 จะให้เป็น high_household

In [44]:
bucket = ['low_household','median_household','high_household']
merge_table3_drop['household_group'] = pd.cut(merge_table3_drop.value_household,
                            bins=[0,0.1,0.3,0.6],
                            labels=bucket,
                            right=False).astype(str).astype(CategoricalDtype(bucket,ordered=True))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


column ชื่อว่า value_population

ถ้ามีค่าตั้งแต่ 0 - 0.05 จะให้เป็น low_population

ถ้ามีค่าตั้งแต่ 0.05 - 0.10 จะให้เป็น median_population

ถ้ามีค่าตั้งแต่ 0.10 - 0.18 จะให้เป็น high_population

In [45]:
bucket = ['low_population','median_population','high_population']
merge_table3_drop['population_group'] = pd.cut(merge_table3_drop.value_population,
                            bins=[0,0.05,0.10,0.18],
                            labels=bucket,
                            right=False).astype(str).astype(CategoricalDtype(bucket,ordered=True))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


column ชื่อว่า value_mobile

ถ้ามีค่าตั้งแต่ 0 - 2000000 จะให้เป็น low_mobile

ถ้ามีค่าตั้งแต่ 2000000 - 5000000 จะให้เป็น median_mobile

ถ้ามีค่าตั้งแต่ 5000000 - 12000000 จะให้เป็น high_mobile

In [47]:
bucket = ['low_mobile','median_mobile','high_mobile']
merge_table3_drop['mobile_group'] = pd.cut(merge_table3_drop.value_mobile,
                            bins=[0,50000000,100000000,140000000],
                            labels=bucket,
                            right=False).astype(str).astype(CategoricalDtype(bucket,ordered=True))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


ตารางแสดงผลจากการที่เราเพิ่ม column

In [48]:
merge_table3_drop

no. Mobile Market Report  ...  population_group  mobile_group
0     5   Mobile Subscribers  ...    low_population    low_mobile
1     6   Mobile Subscribers  ...    low_population    low_mobile
2     7   Mobile Subscribers  ...    low_population    low_mobile
3     8   Mobile Subscribers  ...    low_population    low_mobile
4     9   Mobile Subscribers  ...    low_population    low_mobile
..  ...                  ...  ...               ...           ...
64   69   Mobile Subscribers  ...   high_population   high_mobile
65   70   Mobile Subscribers  ...   high_population   high_mobile
66   71   Mobile Subscribers  ...   high_population   high_mobile
67   72   Mobile Subscribers  ...   high_population   high_mobile
68   73   Mobile Subscribers  ...   high_population   high_mobile

[69 rows x 15 columns]

ตัดตารางให้เหลือแต่กลุ่มของ column ที่ต้องการใช้

In [49]:
merge_table3_drop.columns

Index(['no.', 'Mobile Market Report', 'value_mobile', 'quarter',
       'Thailand Internet Users_subscriber', 'year', 'value_subscriber',
       'type_population', 'value_population', 'type_household',
       'value_household', 'subscriber_group', 'household_group',
       'population_group', 'mobile_group'],
      dtype='object')

In [50]:
data_cut = merge_table3_drop[['subscriber_group', 'household_group', 'population_group', 'mobile_group']]
data_cut

subscriber_group household_group population_group mobile_group
0    low_subscriber   low_household   low_population   low_mobile
1    low_subscriber   low_household   low_population   low_mobile
2    low_subscriber   low_household   low_population   low_mobile
3    low_subscriber   low_household   low_population   low_mobile
4    low_subscriber   low_household   low_population   low_mobile
..              ...             ...              ...          ...
64  high_subscriber  high_household  high_population  high_mobile
65  high_subscriber  high_household  high_population  high_mobile
66  high_subscriber  high_household  high_population  high_mobile
67  high_subscriber  high_household  high_population  high_mobile
68  high_subscriber  high_household  high_population  high_mobile

[69 rows x 4 columns]

In [ ]:
data_cut 

**Assosiation ตาราง data_cut**

Transection คือ ข้อมูล Internet ความเร็วสูง

item คือ ระดับ low median high

ค่าสนับสนุน (Support) คือ เปอร์เซ็นต์ของจำนวน Item sets ทั้งหมดที่เกิดขึ้นในฐานข้อมูล

ค่าความเชื่อมั่น (Confidence) คือ เปอร์เซ็นต์ของจำนวน Item sets ทั้งหมดที่เกิดขึ้นในฐานข้อมูลต่อจำนวน Item sets ที่เกิดขึ้น

ค่าสหสัมพันธ์ (Lift) คือค่าที่บ่งบอกว่ากฎมีความสัมพันธ์กัน หรือไม่ โดยถ้าLift มีค่ามากกว่า 1 แสดงว่ากฎทั้งสองมีความสัมพันธ์กัน แต่ถ้ามีค่าน้อยกว่า 1 แสดงว่า กฎไม่ความสัมพันธ์กัน

In [51]:
transactions = data_cut

##Apriori

จะไปหาละก็คำนวณ Support ให้เรา

http://zaxrosenberg.com/unofficial-apyori-documentation/

In [52]:
!pip install apyori

In [53]:
from apyori import apriori

In [59]:
transacs = []   ## กำหนด transacs ให้เป็น list ว่าง
for i in range(0,len(data_cut)):
  transacs.append([str(data_cut.values[i,j]) for j in range(0,4)]) 

In [60]:
Association = list(apriori(transacs,min_support = 0.40, min_cofidence = 0.50))
Association

[RelationRecord(items=frozenset({'high_subscriber'}), support=0.42028985507246375, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'high_subscriber'}), confidence=0.42028985507246375, lift=1.0)]),
 RelationRecord(items=frozenset({'low_population'}), support=0.4057971014492754, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'low_population'}), confidence=0.4057971014492754, lift=1.0)]),
 RelationRecord(items=frozenset({'median_household'}), support=0.43478260869565216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'median_household'}), confidence=0.43478260869565216, lift=1.0)]),
 RelationRecord(items=frozenset({'median_mobile'}), support=0.4492753623188406, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'median_mobile'}), confidence=0.4492753623188406, lift=1.0)]),
 RelationRecord(items=frozenset({'median_mobile', 'median_household'}), support=0.43

#Visualization

In [61]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [62]:
X = data_cut.values
X

array([['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mobile'],
       ['low_subscriber', 'low_household', 'low_population',
        'low_mo

In [58]:
data_cut.columns

Index(['subscriber_group', 'household_group', 'population_group',
       'mobile_group'],
      dtype='object')

In [42]:
data_cut['subscriber_group'] = data_cut['subscriber_group'].map({'low':0,'median':1,'high':2})
data_cut['household_group'] = data_cut['household_group'].map({'low':0,'median':1,'high':2})
data_cut['population_group'] = data_cut['population_group'].map({'low':0,'median':1,'high':2})
data_cut['mobile_group'] = data_cut['mobile_group'].map({'low':0,'median':1,'high':2})
data_cut

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

subscriber_group household_group population_group mobile_group
0                 0               0                0            0
1                 0               0                0            0
2                 0               0                0            0
3                 0               0                0            0
4                 0               0                0            0
..              ...             ...              ...          ...
64                2               2                2            2
65                2               2                2            2
66                2               2                2            2
67                2               2                2            2
68                2               2                2            2

[69 rows x 4 columns]